File: app/src/main/java/info/maaskant/wouttest2/navigation/NodeListAdapter.java
package info.maaskant.wouttest2.navigation;

import static android.view.View.OnClickListener;

import java.util.ArrayList;
import java.util.List;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import info.maaskant.wouttest2.R;
import info.maaskant.wouttest2.model.ContentNode;
import info.maaskant.wouttest2.model.FolderNode;
import info.maaskant.wouttest2.model.Node;
import info.maaskant.wouttest2.model.NodeVisitor;
import timber.log.Timber;

class NodeListAdapter extends RecyclerView.Adapter<NodeListAdapter.NodeViewHolder> {

    private final List<Node> nodes = new ArrayList<>();

    private OnClickListener onClickListener;

    public NodeListAdapter(List<Node> nodes) {
        setHasStableIds(true);
        this.nodes.addAll(nodes);
    }

    public void setOnClickListener(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    public Node getItem(int position) {
        return nodes.get(position);
    }

    @Override
    public NodeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.node_list_item, parent,
                false);
        return new NodeViewHolder(v, onClickListener);
    }

    @Override
    public void onBindViewHolder(NodeViewHolder holder, int position) {
        Node node = nodes.get(position);
        holder.titleTextView.setText(node.getName());
        node.accept(new NodeVisitor() {
            @Override
            public void visit(@NonNull ContentNode node) {
                holder.iconImageView.setImageResource(R.drawable.ic_insert_drive_file_black_24dp);
            }

            @Override
            public void visit(@NonNull FolderNode node) {
                holder.iconImageView.setImageResource(R.drawable.ic_folder_black_24dp);
            }
        });
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return nodes.size();
    }

    public void set(List<Node> nodes) {
        this.nodes.clear();
        this.nodes.addAll(nodes);

        notifyDataSetChanged();
    }

    public static class NodeViewHolder extends RecyclerView.ViewHolder {

        @NonNull
        public final ImageView iconImageView;

        @NonNull
        public final TextView titleTextView;

        public NodeViewHolder(View view, OnClickListener onClickListener) {
            super(view);
            iconImageView = (ImageView) view.findViewById(R.id.icon);
            titleTextView = (TextView) view.findViewById(R.id.title);
            view.setOnClickListener(onClickListener);
        }
    }
}

File: app/src/main/java/info/maaskant/wmsnotes/android/ui/navigation/NodeListAdapter.kt
package info.maaskant.wmsnotes.android.ui.navigation

import android.view.HapticFeedbackConstants
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.google_material_typeface_library.GoogleMaterial
import com.mikepenz.iconics.IconicsDrawable
import info.maaskant.wmsnotes.R
import info.maaskant.wmsnotes.client.indexing.Folder
import info.maaskant.wmsnotes.client.indexing.Node
import info.maaskant.wmsnotes.client.indexing.Note
import kotlin.properties.Delegates
import kotlin.random.Random

internal class NodeListAdapter(
    private val listener: NodeListAdapterListener
) :
    RecyclerView.Adapter<NodeListAdapter.NodeViewHolder>(),
    AutoUpdatableAdapter {

    // Source: https://github.com/antoniolg/diffutil-recyclerview-kotlin/blob/master/app/src/main/java/com/antonioleiva/diffutilkotlin/ContentAdapter.kt
    var items: List<Node> by Delegates.observable(emptyList()) { _, oldList, newList ->
        autoNotify(oldList, newList) { o, n -> o.aggId == n.aggId }
    }

    init {
        setHasStableIds(true)
        this.items = emptyList()
    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun getItemId(position: Int): Long {
        // This is not really correct, since two different objects may have the same hash code. However, we are going
        // to rely on our implementation in Note/Folder.hashCode()
        return items[position].hashCode().toLong()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NodeViewHolder {
        val view =
            LayoutInflater.from(parent.context).inflate(R.layout.node_list_item, parent, false)
        return NodeViewHolder(view, listener)
    }

    override fun onBindViewHolder(holder: NodeViewHolder, position: Int) {
        holder.bind(items[position])
    }

    interface NodeListAdapterListener {
        fun onClick(node: Node)
        fun onLongClick(node: Node)
    }

    class NodeViewHolder(
        view: View,
        private val listener: NodeListAdapterListener
    ) :
        RecyclerView.ViewHolder(view),
        View.OnClickListener,
        View.OnLongClickListener {
        private lateinit var node: Node
        private val main: LinearLayout = view.findViewById(R.id.node_item)
        private val icon: ImageView = view.findViewById(R.id.icon)
        private val title: TextView = view.findViewById(R.id.title)

        init {
            view.setOnClickListener(this)
            view.setOnLongClickListener(this)
        }

        fun bind(node: Node) {
            this.node = node
            main.isActivated = Random.nextBoolean()
            title.text = node.title
            when (node) {
                is Note -> {
                    icon.setImageDrawable(
                        IconicsDrawable(
                            icon.context,
                            GoogleMaterial.Icon.gmd_insert_drive_file
                        )
                    )
                }
                is Folder -> {
                    icon.setImageDrawable(
                        IconicsDrawable(
                            icon.context,
                            GoogleMaterial.Icon.gmd_folder
                        )
                    )
                }
            }
        }

        override fun onClick(view: View) {
            listener.onClick(node)
        }

        override fun onLongClick(view: View): Boolean {
            view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            listener.onLongClick(node)
            return true
        }
    }
}
